File: .config/ags/widget/desktop/Mapkey.ts
import PopupWindow from "widget/PopupWindow";
import { sh } from "lib/utils";
import icons from "lib/icons"

function CommandOutput() {
const commandOutput = Utils.exec(`
awk '
    BEGIN {
        FS = " = "; 
        max_columns = 2;  
        column_count = 0;
    }

    /^bind/ || /^bindm/ {
        gsub(/\$/, "", $0);

        keybinding = $2;
        action = $3;
        gsub(/^bind/, "", keybinding);
        gsub(/SUPER,/, "SUPER+", keybinding);  
        gsub(/SHIFT,/, "SHIFT+", keybinding);

        gsub(/,/, " ", keybinding);  
        gsub(/,/, " ", action);    
    
        gsub(/_/, "+", keybinding);  
        gsub(/_/, "+", action);    

        printf "%-30s %-40s", keybinding, action;
        column_count++; 
 
        if (column_count == max_columns) {
            print "";
            column_count = 0;
        }
    }

    END {
        if (column_count > 0) {
            print "";  
        }
    }
' /home/shen/.config/hypr/rules/keybinds.conf
`);

  return Widget.Box({
    className: 'key',
    vertical: true,
    spacing: 6,
    children: commandOutput.split("\n").map(line => Widget.Label({ label: line })),
  });
}

export default () => PopupWindow({
  name: "Mapkey",
  layout: "center",
  child: Widget.Box({
    className: 'mapkey',
    vertical: false,
    children: [
      Widget.Box({
        vertical: true, 
        children: [
          Widget.Box({
            vertical: false,
            children: [
              Widget.Icon({
                icon: icons.hypr.hypr,  
                css: 'margin-right: 5px;',     
              }),
              Widget.Label({
                label: "Keymap Hyprland",
                className: 'title',
                css: 'font-size: 15px;',
              }),
            ],
          }),
          CommandOutput(),
        ],
      }),
    ],
  }),
});

File: .config/ignis/modules/mapkey/mapkey.py
from ignis.widgets import Widget
import subprocess


def get_keybindings() -> Widget.Box:
    command = """
        awk '
        BEGIN {
            FS = "[ ]*=[ ]*";
        }

        /^bind/ || /^bindm/ {
            keybinding = $2;
            action = $3;

            gsub(/SUPER,/, "SUPER+", keybinding);
            gsub(/SHIFT,/, "SHIFT+", keybinding);
            gsub(/CTRL,/, "CTRL+", keybinding);
            gsub(/0x002d/, "11", keybinding);
            gsub(/0x003d/, "12", keybinding);

            gsub(/,/, " ", keybinding);
            gsub(/,/, " ", action);
            gsub(/_/, "+", keybinding);
            gsub(/_/, "+", action);

            print keybinding, action;
        } 
        ' /home/shen/.config/hypr/rules/keybinds.conf
    """

    result = subprocess.run(command, shell=True, capture_output=True, text=True)
    output = result.stdout.strip()

    lines = output.split("\n")
    columns = [[], []]

    for idx, line in enumerate(lines):
        if line.strip() != "":
            parts = line.split()
            if len(parts) >= 2:  
                keybinding = parts[0]
                action = " ".join(parts[1:])  
                label = Widget.Label(label=f"{keybinding} {action}")
                separator = Widget.Separator(css_classes=["separator-mp"])

                column_idx = idx % 2
                columns[column_idx].append(label)
                columns[column_idx].append(separator)

    children = []
    for column in columns:
        children.append(Widget.Box(vertical=True, child=column))

    return Widget.Box(
        vertical=False,
        child=children
    )
